src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs

```diff
@@ -360,11 +380,29 @@ public IEnumerable<StatementSyntax> GenerateGuaranteedUnmarshalStatements(StubId
             (string managedIdentifier, string nativeIdentifier) = context.GetIdentifiers(TypeInfo);
             string numElementsIdentifier = MarshallerHelpers.GetNumElementsIdentifier(TypeInfo, context);
 
+            // Generate numElements expression with null check for pointer types
+            ExpressionSyntax numElementsExpression = ElementsMarshalling.GenerateNumElementsExpression(countInfo, countInfoRequiresCast, CodeContext, context);
+
+            // If the marshalling direction is unmanaged-to-managed and we have a native pointer type,
+            // we need to check if the native pointer is null before using the size parameter to avoid
+            // allocating arrays for null pointers.
+            if (CodeContext.Direction == MarshalDirection.UnmanagedToManaged &&
```

Comment:
Change the direction checks in your changes to check for "not ManagedToUnmanaged" and add some tests for arrays passed by `ref`